stats/opencensus/stats.go

```diff
@@ -0,0 +1,227 @@
+	return context.WithValue(ctx, rpcDataKey{}, d)
+}
+
+var (
+	// bounds separate variable for testing purposes.
+	bytesDistributionBounds  = []float64{1024, 2048, 4096, 16384, 65536, 262144, 1048576, 4194304, 16777216, 67108864, 268435456, 1073741824, 4294967296}
+	bytesDistribution        = view.Distribution(bytesDistributionBounds...)
+	millisecondsDistribution = view.Distribution(0.01, 0.05, 0.1, 0.3, 0.6, 0.8, 1, 2, 3, 4, 5, 6, 8, 10, 13, 16, 20, 25, 30, 40, 50, 65, 80, 100, 130, 160, 200, 250, 300, 400, 500, 650, 800, 1000, 2000, 5000, 10000, 20000, 50000, 100000)
+	countDistributionBounds  = []float64{1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768, 65536} // float64 correct here?
+	countDistribution        = view.Distribution(countDistributionBounds...)
+)
+
+func removeLeadingSlash(mn string) string {
+	return strings.TrimLeft(mn, "/")
+}
+
+// rpcData is data about the rpc attempt client side, and the overall rpc server
+// side.
+type rpcData struct {
+	// access these counts atomically for hedging in the future
+	// number of messages sent from side (client || server)
+	sentMsgs int64
+	// number of bytes sent (within each message) from side (client || server)
+	sentBytes int64
+	// number of messages received on side (client || server)
+	recvMsgs int64
+	// number of bytes received (within each message) received on side (client
+	// || server)
+	recvBytes int64
+
+	startTime time.Time // is this still the correct way to count latency?
+	method    string
+}
+
+// statsTagRPC creates a recording object to derive measurements from in the
+// context, scoping the recordings to per RPC Attempt client side (scope of the
+// context). It also populates the gRPC Metadata within the context with any
+// opencensus specific tags set by the application in the context, binary
+// encoded to send across the wire.
+func (csh *clientStatsHandler) statsTagRPC(ctx context.Context, info *stats.RPCTagInfo) context.Context {
+	d := &rpcData{
+		startTime: time.Now(),
+		method:    info.FullMethodName,
+	}
+	// Populate gRPC Metadata with OpenCensus tag map if set by application.
+	if tm := tag.FromContext(ctx); tm != nil {
+		ctx = stats.SetTags(ctx, tag.Encode(tm))
+	}
+	return setRPCData(ctx, d)
+}
+
+// statsTagRPC creates a recording object to derive measurements from in the
+// context, scoping the recordings to per RPC server side (scope of the
+// context). It also deserializes the opencensus tags set in the context's gRPC
+// Metadata, and adds a server method tag to the opencensus tags.
+func (ssh *serverStatsHandler) statsTagRPC(ctx context.Context, info *stats.RPCTagInfo) context.Context {
+	d := &rpcData{
+		startTime: time.Now(),
+		method:    info.FullMethodName,
+	}
+	if tagsBin := stats.Tags(ctx); tagsBin != nil {
+		if tags, err := tag.Decode(tagsBin); err != nil {
+			ctx = tag.NewContext(ctx, tags)
+		}
+	}
+	ctx, _ = tag.New(ctx, tag.Upsert(keyServerMethod, removeLeadingSlash(info.FullMethodName)))
```

Comment:
Is it safe to ignore this error? Why?

Reply:
> If the error does hit, it means that the tag map does not have the correct KeyServerMethod. This isn't used for any metrics, and is really just populated for the caller to pull out for their own use

If it returns an error, is the context it returns valid?  Or does it return `nil`?  If the latter, then we need to check the error just in case.

Reply:
Discussed offline; wrote comment explaining the thought process here.